### ISSUE TITLE
Use sigstore tools to sign artifacts and distribute them through an OCI registry

### ISSUE THREAD

**Issue by charithe** [cerbos/cerbos#239]
[Sigstore](https://www.sigstore.dev) is a suite of tools for software supply chain security. [Cosign](https://github.com/sigstore/cosign) the signing tool is now stable and can be used to sign and distribute containers AND binary blobs using an OCI registry (both GitHub and JFrog are supported).

The [keyless signing mode](https://github.com/sigstore/cosign/blob/main/KEYLESS.md) is still experimental though. When it becomes stable, we should configure CI to sign all artifacts during the release. 

To do:

- [ ] Wait for keyless mode to be stable (because managing keys manually with a KMS is a bit of a faff)
- [ ] Pick a Google account from @cerbos.dev  to be the signer
- [ ] Configure CI to sign containers, binaries and any other artifacts
- [ ] Use the same OCI registry as the containers to distribute binaries as well (see https://oras.land)  

```sh
COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/cerbos/cerbos:0.4.0

Verification for ghcr.io/cerbos/cerbos:0.4.0 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - Any certificates were verified against the Fulcio roots.

Certificate subject:  [charith@cerbos.dev]
{"critical":{"identity":{"docker-reference":"ghcr.io/cerbos/cerbos"},"image":{"docker-manifest-digest":"sha256:7d9e3a7766c1f94288d11a2f3490c1c19f5b9e469027e4cb3394fe6a02d45049"},"type":"cosign container image signature"},"optional":null}
```

**Comment by charithe** [comment]
Cosign has experimental support for signing using the GitHub OIDC -- which means we don't have to manage a private key separately. Unfortunately, as of 08-11-2021, the Rekor transparency log is not fully functional:

```
Error: signing dist/actions-test_0.1.0-prerelease_Darwin_all.tar.gz: Post "https://rekor.sigstore.dev/api/v1/log/entries": context deadline exceeded
main.go:48: error during command execution: signing dist/actions-test_0.1.0-prerelease_Darwin_all.tar.gz: Post "https://rekor.sigstore.dev/api/v1/log/entries": context deadline exceeded
```

Signing config:

```yaml
# GH Actions
      - uses: sigstore/cosign-installer@main

      - name: GoReleaser
        uses: goreleaser/goreleaser-action@v2
        with:
          version: latest
          args: release --config=.goreleaser-ci.yml --rm-dist
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          COSIGN_EXPERIMENTAL: "1"
```

```yaml
# Goreleaser
   signs:
     - id: cosign
       cmd: cosign
       args: ["sign-blob", "-output=${signature}", "${artifact}"]
       artifacts: all
```

**Comment by charithe** [comment]
There is an official GitHub action for this now: https://github.com/slsa-framework/slsa-github-generator-go

https://github.blog/2022-04-07-slsa-3-compliance-with-github-actions/

**Comment by psolarcz** [comment]
@charithe 
I'd like to help with this but not sure if I'll be able to test it.

I have example for signing docker container in my [repo](https://github.com/psolarcz/sign-images-keyless) - it should be possible with goreleaser too.

**Comment by charithe** [comment]
Thanks for the offer and we appreciate your help. I think this one has to be done by someone from the Cerbos team because it requires access to GitHub settings etc.

**Comment by psolarcz** [comment]
@charithe no extra settings are required.
Managed to do POC for docker container with goreleaser:

``` bash
cosign verify  --rekor-url "https://rekor.sigstore.dev/" ttl.sh/signed-test-a161b2c:1h  --certificate-identity "https://github.com/psolarcz/sign-images-keyless/.github/workflows/release.yml@refs/tags/v6" --certificate-oidc-issuer "https://token.actions.githubusercontent.com"

Verification for ttl.sh/signed-test-a161b2c:1h --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The code-signing certificate was verified using trusted certificate authority certificates

[{"critical":{"identity":{"docker-reference":"ttl.sh/signed-test-a161b2c"},"image":{"docker-manifest-digest":"sha256:8cf5034386b5e3a3a4973081847a51408eb842ac89345418cfafc7d7a830cf89"},"type":"cosign container image signature"},"optional":{"1.3.6.1.4.1.57264.1.1":"https://token.actions.githubusercontent.com","1.3.6.1.4.1.57264.1.2":"push","1.3.6.1.4.1.57264.1.3":"a161b2c1996d6cd78f89ce74522215c9d00c9c1b","1.3.6.1.4.1.57264.1.4":"goreleaser","1.3.6.1.4.1.57264.1.5":"psolarcz/sign-images-keyless","1.3.6.1.4.1.57264.1.6":"refs/tags/v6","Bundle":{"SignedEntryTimestamp":"MEYCIQClVwiJe8ceAaJ7y5fX1qD457GUkffjJ54eCavCXBzb8AIhAPIjgKaNOW3ocznvxuocRn9KbPnevbDhcb4vdu4IV3jS","Payload":{"body":"eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJhZDljODZmNDBlODI3NmM2NDc2OTVhYmE0MjI2NWQ0MjE2NGQwNjcxNDk0NGZkNzMyODg3YWZiMDU1NmFkZDM3In19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FUUNJQ0I3c3FVaEhGR05XQjltTEtyNEM3OTMya0YzZWZEbDI5MHZYOEJ0VDJIRUFpQW9OdUd0WTlYS0dxOGt5SHVHeFFOTEFxaE5lYk5YQWFLK0hNemhLN3FTNGc9PSIsInB1YmxpY0tleSI6eyJjb250ZW50IjoiTFMwdExTMUNSVWRKVGlCRFJWSlVTVVpKUTBGVVJTMHRMUzB0Q2sxSlNVY3lha05EUW0xRFowRjNTVUpCWjBsVlJrZFVkSE5HYWxGSVJtUTRhazFKYXpGVlltWjZOVlppWkZWcmQwTm5XVWxMYjFwSmVtb3dSVUYzVFhjS1RucEZWazFDVFVkQk1WVkZRMmhOVFdNeWJHNWpNMUoyWTIxVmRWcEhWakpOVWpSM1NFRlpSRlpSVVVSRmVGWjZZVmRrZW1SSE9YbGFVekZ3WW01U2JBcGpiVEZzV2tkc2FHUkhWWGRJYUdOT1RXcFJkMDFVU1RCTlZGa3dUbnBWTkZkb1kwNU5hbEYzVFZSSk1FMVVXVEZPZWxVMFYycEJRVTFHYTNkRmQxbElDa3R2V2tsNmFqQkRRVkZaU1V0dldrbDZhakJFUVZGalJGRm5RVVUyVFdsSFdIY3ZOemxrVTNjcldGaFZZMEZNYW5wU2VHWTNMMnhFVEVWMmVEaERLMVlLUWxGWUwwUkJkR1Z5TW5aTGNYSlBVekJ2WkZJelNXMHhPVzE2UTB4WU5VczBWaXRJYTFORmFscDJOMVJZTTNCV2RFdFBRMEpZT0hkbloxWTNUVUUwUndwQk1WVmtSSGRGUWk5M1VVVkJkMGxJWjBSQlZFSm5UbFpJVTFWRlJFUkJTMEpuWjNKQ1owVkdRbEZqUkVGNlFXUkNaMDVXU0ZFMFJVWm5VVlZxVEVGc0NtMXlhMUl6U2lzelYxZEtXRmRFSzJFM1pFRndhME5aZDBoM1dVUldVakJxUWtKbmQwWnZRVlV6T1ZCd2VqRlphMFZhWWpWeFRtcHdTMFpYYVhocE5Ga0tXa1E0ZDJGQldVUldVakJTUVZGSUwwSkdOSGRZU1ZwaFlVaFNNR05JVFRaTWVUbHVZVmhTYjJSWFNYVlpNamwwVEROQ2VtSXllR2hqYlU0MlRETk9jQXBhTWpSMFlWY3hhRm95Vm5wTVYzUnNaVmQ0YkdNelRYWk1iV1J3WkVkb01WbHBPVE5pTTBweVdtMTRkbVF6VFhaamJWWnpXbGRHZWxwVE5UVmlWM2hCQ21OdFZtMWplVGt3V1Zka2Vrd3pXVEpOUkd0SFEybHpSMEZSVVVKbk56aDNRVkZGUlVzeWFEQmtTRUo2VDJrNGRtUkhPWEphVnpSMVdWZE9NR0ZYT1hVS1kzazFibUZZVW05a1Ywb3hZekpXZVZreU9YVmtSMVoxWkVNMWFtSXlNSGRGWjFsTFMzZFpRa0pCUjBSMmVrRkNRV2RSUldOSVZucGhSRUV5UW1kdmNncENaMFZGUVZsUEwwMUJSVVJDUTJob1RWUlplRmxxU21wTlZHczFUbTFSTWxreVVUTlBSMWswVDFkT2JFNTZVVEZOYWtsNVRWUldhazlYVVhkTlIwMDFDbGw2Um1sTlFtZEhRMmx6UjBGUlVVSm5OemgzUVZGUlJVTnRaSFpqYlZaeldsZEdlbHBZU1hkTFoxbExTM2RaUWtKQlIwUjJla0ZDUWxGUlkyTklUbllLWWtkR2VWa3piM1pqTW14dVlta3hjR0pYUm01YVdFMTBZVEpXTldKSFZucGpla0ZoUW1kdmNrSm5SVVZCV1U4dlRVRkZSMEpCZUhsYVYxcDZURE5TYUFwYU0wMTJaR3BaZDA5M1dVdExkMWxDUWtGSFJIWjZRVUpEUVZGMFJFTjBiMlJJVW5kamVtOTJURE5TZG1FeVZuVk1iVVpxWkVkc2RtSnVUWFZhTW13d0NtRklWbWxrV0U1c1kyMU9kbUp1VW14aWJsRjFXVEk1ZEUxSGIwZERhWE5IUVZGUlFtYzNPSGRCVVd0RldFRjRZV0ZJVWpCalNFMDJUSGs1Ym1GWVVtOEtaRmRKZFZreU9YUk1NMEo2WWpKNGFHTnRUalpNTTA1d1dqSTBkR0ZYTVdoYU1sWjZURmQwYkdWWGVHeGpNMDEyVEcxa2NHUkhhREZaYVRrellqTktjZ3BhYlhoMlpETk5kbU50Vm5OYVYwWjZXbE0xTldKWGVFRmpiVlp0WTNrNU1GbFhaSHBNTTFreVRVUm5SME5wYzBkQlVWRkNaemM0ZDBGUmIwVkxaM2R2Q2xsVVJUSk5WMGw1V1hwRk5VOVVXbXRPYlU1clRucG9iVTlFYkdwYVZHTXdUbFJKZVUxcVJURlplbXhyVFVSQ2FrOVhUWGhaYWtGa1FtZHZja0puUlVVS1FWbFBMMDFCUlV4Q1FUaE5SRmRrY0dSSGFERlphVEZ2WWpOT01GcFhVWGRRZDFsTFMzZFpRa0pCUjBSMmVrRkNSRUZSZUVSRE9XOWtTRkozWTNwdmRncE1NbVJ3WkVkb01WbHBOV3BpTWpCMlkwaE9kbUpIUm5sWk0yOTJZekpzYm1KcE1YQmlWMFp1V2xoTmRHRXlWalZpUjFaNlkzcEJORUpuYjNKQ1owVkZDa0ZaVHk5TlFVVk9Ra052VFV0SFJYaE9ha1pwVFcxTmVFOVVhekphUkZwcVdrUmpORnBxWnpWWk1sVXpUa1JWZVUxcVNYaE9WMDAxV2tSQmQxbDZiR29LVFZkSmQwaEJXVXRMZDFsQ1FrRkhSSFo2UVVKRVoxRlBSRUY0ZVZwWFducE1NMUpvV2pOTmRtUnFXWGRIVVZsTFMzZFpRa0pCUjBSMmVrRkNSSGRSVEFwRVFXc3pUa1JqTTAxNlozaE9ha1YzUzNkWlMwdDNXVUpDUVVkRWRucEJRa1ZCVVdSRVFuUnZaRWhTZDJONmIzWk1NbVJ3WkVkb01WbHBOV3BpTWpCMkNtTklUblppUjBaNVdUTnZkMGRCV1V0TGQxbENRa0ZIUkhaNlFVSkZVVkZMUkVGbmVrMTZaelJOUkVVMFRucENjVUpuYjNKQ1owVkZRVmxQTDAxQlJWTUtRa1ozVFZkdGFEQmtTRUo2VDJrNGRsb3liREJoU0ZacFRHMU9kbUpUT1hkak1qbHpXVmhLYW1WcE9YcGhWMlIxVEZkc2RGbFhaR3hqZVRGeVdsaHNjd3BhV0U1NlRIazFibUZZVW05a1YwbDJaREk1ZVdFeVduTmlNMlI2VEROS2JHSkhWbWhqTWxWMVpWY3hjMUZJU214YWJrMTJaRWRHYm1ONU9USk9ha0UwQ2tKbmIzSkNaMFZGUVZsUEwwMUJSVlJDUTI5TlMwZEZlRTVxUm1sTmJVMTRUMVJyTWxwRVdtcGFSR00wV21wbk5Wa3lWVE5PUkZWNVRXcEplRTVYVFRVS1drUkJkMWw2YkdwTlYwbDNSa0ZaUzB0M1dVSkNRVWRFZG5wQlFrWkJVVWRFUVZKM1pGaE9iMDFIU1VkRGFYTkhRVkZSUW1jM09IZEJVbFZGVmtGNFV3cGhTRkl3WTBoTk5reDVPVzVoV0ZKdlpGZEpkVmt5T1hSTU0wSjZZako0YUdOdFRqWk1NMDV3V2pJMGRHRlhNV2hhTWxaNlRGZDBiR1ZYZUd4ak0wMTJDbGxYVGpCaFZ6bDFZM2s1ZVdSWE5YcE1lbU15VGtSTk1FNUVXVEJOVkZWMldWaFNNRnBYTVhka1NFMTJUVlJCVjBKbmIzSkNaMFZGUVZsUEwwMUJSVmNLUWtGblRVSnVRakZaYlhod1dYcERRbWwzV1V0TGQxbENRa0ZJVjJWUlNVVkJaMUk1UWtoelFXVlJRak5CVGpBNVRVZHlSM2g0UlhsWmVHdGxTRXBzYmdwT2QwdHBVMncyTkROcWVYUXZOR1ZMWTI5QmRrdGxOazlCUVVGQ2FsUjRaa040WTBGQlFWRkVRVVZuZDFKblNXaEJUSFpQYURocVExSTNiRkZ0UjBzeENsZEtVVk5YYzFOcVFscDBUMjlEYm1OVFV6bHhhbmN3YlRSck5rbEJhVVZCTWt3dlNEQXdTMFpwTkhweU1GZHphRzVDU25kUE9VeG5jbGx3ZUN0VlVuWUtaekZTYkhaU01TdHJiVVYzUTJkWlNVdHZXa2w2YWpCRlFYZE5SR0ZCUVhkYVVVbDNaR3cxTm1kV2RXeG1URTExVTFndlExVlRaV1JCT1RacGIwVjFTZ3BqY0VKek5tNXNkMVZEUkd4TFkwcEVaa3hQWTBOc1lWRmtiMGxqVEM5RFNVRnVXbFZCYWtWQk0xRkRWVkV2U0ZoVksydFJNRmRXWVRCdlJYSjFkbVV3Q2tGd1R6ZHRVSE5YZWxCdWFUZFJSbmxtSzBzeVRFWm1jMUpYVkRGSmJUVjNkMGxYTUM5ak5rd0tMUzB0TFMxRlRrUWdRMFZTVkVsR1NVTkJWRVV0TFMwdExRbz0ifX19fQ==","integratedTime":1706114878,"logIndex":66155399,"logID":"c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d"}},"Issuer":"https://token.actions.githubusercontent.com","Subject":"https://github.com/psolarcz/sign-images-keyless/.github/workflows/release.yml@refs/tags/v6","githubWorkflowName":"goreleaser","githubWorkflowRef":"refs/tags/v6","githubWorkflowRepository":"psolarcz/sign-images-keyless","githubWorkflowSha":"a161b2c1996d6cd78f89ce74522215c9d00c9c1b","githubWorkflowTrigger":"push"}}]
```

I can create a PR but will not be able to test, so I'll need some help from your side.

**Comment by charithe** [comment]
Fixed in #1959. Thanks @psolarcz
